비밀번호 찾기

kihyuk-jeong opened the pull request:
1. 찾고자 하는 계정의 아이디(이메일)을 입력
2. 등록된 이메일 또는 휴대폰으로 인증번호 확인
3. 인증번호 확인 완료시 비밀번호 변경 진행

<issue-number>
 #15 : 비밀번호 찾기 구현

kihyuk-jeong committed:
feature : 비밀번호 찾기 구현

1. 찾고자 하는 계정의 아이디(이메일)을 입력
2. 등록된 이메일 또는 휴대폰으로 인증번호 확인
3. 인증번호 확인 완료시 비밀번호 변경 진행

<issue-number>
 #15 : 비밀번호 찾기 구현

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/common/util/coolSms/coolSmsConstants.java:
```diff
@@ -1,10 +1,10 @@
 public class coolSmsConstants {
-    public static String COOLSMS_KEY = "NCSNEETJ1Z3IPXZ0";
-    public static String COOLSMS_SECRET = "1SBBED1RWKCUJBD46AYDJT1IDSP0PHQO";
+    public static String COOLSMS_KEY = "NCSBVON9BZGTHH2U";
+    public static String COOLSMS_SECRET = "CLFRQNF13AWV5VHXSMTFWZKIZRJISLK7";
```
해당 키값은 테스트용으로 사용 후 추후 외부주입으로 변경하겠습니다.!

f-lab-dev replied:
네 아래에 핸드폰번호도 있던데 이런 개인정보가 커밋내역을 통해 노출되니 조심하셔야합니다~

Hosick commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -86,5 +93,32 @@ public ResponseEntity logout() {
+    //휴대폰 인증 선택시 : sendSms / SmsVerification 핸들러
+    //이메일 인증 선택시 : sendEmail / emailVerification 핸들러
+    @GetMapping("/find/password/{email}")
+    public ResponseEntity<FindUserRequest> forget_password(@PathVariable String email) {
```
몇몇 메소드의 네이밍을 `snake case`로 정하셨던데 혹시 이유를 알 수 있을까요??

kihyuk-jeong replied:
메서드 이름을 행위에 맞게 전달하려는 과정에서 네이밍 했는데 `snake case`로 정의하면 메소드 이름의 통일성이 깨지는 것 같네요. 수정하겠습니다!

Hosick commented on src/main/java/com/flab/shoeauction/common/util/email/EmailConstants.java:
```diff
@@ -0,0 +1,8 @@
+
+public class EmailConstants {
+
+    public static final String TITLE = "Shoe-action 인증번호 안내";
```
action 오타인 것 같습니다!

Hosick commented on src/main/java/com/flab/shoeauction/common/util/response/ResponseConstants.java:
```diff
@@ -22,7 +22,7 @@
+            "가입하지 않은 아이디이거나, 잘못된 비밀번호입니다.", HttpStatus.UNAUTHORIZED
         );
 
     public static final ResponseEntity<String> LOGIN_UNAUTHORIZED =
```
`UNAUTHORIZED`와 `LOGIN_UNAUTHORIZED`의 차이를 알기가 조금 어려운 것 같습니다.
조금만 변수명이 구체적이면 좋을 것 같습니다!

kihyuk-jeong replied:
변수명 구체적으로 수정하겠습니다~

kihyuk-jeong replied:
`UNAUTHORIZED`의 경우, 아이디 또는 비밀번호 불일치(또는 존재하지 않는 아이디)로 로그인에 실패한 경우기 때문에 인증관련 상태코드 보다는 `BAD REQUEST`가 더 나을 것 같아서 변경하겠습니다!

Hosick commented on src/main/java/com/flab/shoeauction/common/util/user/UserConstants.java:
```diff
@@ -1,5 +1,12 @@
+    public static final String makeRandomNumber() {
+        Random random = new Random();
+        return String.valueOf(100000 + random.nextInt(900000));
+    }
```
저는 난수 생성 메소드를 Service에 작성하였는데 @ROMANIAPEOPLE 님은 Constants에 작성하셨네요.
@f-lab-dev 어떤 방법이 더 적절할지 멘토님 의견이 궁금합니다! 아니면 모듈로 빼는 것이 더 좋은 방법일까요?

f-lab-dev replied:
`Constants`라는 이름에서 유추할 수 있듯이 이 클래스의 역할에는 맞지 않아보입니다~

난수 생성은 `Util`클래스로 빼는게 낫지 않을까요? 유틸클래스가 맡는 역할이 무엇인지도 한번 알아보시면 좋겠습니다

kihyuk-jeong replied:
@Hosick  Service 메소드에 작성하게되면 `SmsCertificationService`와 `EmailCertificationService`에 중복 코드가 적용되어서 `Constants`에 일시적으로 작성했습니다.! `Constants`는 상수를 저장하는 공간이므로 `util` 클래스로 따로 빼는게 좋을 것 같습니다!

Hosick commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -86,5 +93,32 @@ public ResponseEntity logout() {
+    @GetMapping("/find/password/{email}")
+    public ResponseEntity<FindUserRequest> forget_password(@PathVariable String email) {
+        FindUserRequest findUserRequest = userService.getPhoneNumber(email);
+        return ResponseEntity.ok(findUserRequest);
```
응답용 Dto면 `findUserResponse`가 맞지 않을까요?

kihyuk-jeong replied:
`findUserResponse`가 맞는 것 같네요 ! 감사합니다. 리뷰 반영하겠습니다~

Hosick commented on src/main/java/com/flab/shoeauction/service/EmailCertificationService.java:
```diff
@@ -0,0 +1,60 @@
+    //인증번호 확인
+    public boolean verifyEmail(EmailCertificationRequest requestDto){
+        if (emailCertificationDao.hasKey(requestDto.getEmail()) &&
+            emailCertificationDao.getEmailCertification(requestDto.getEmail()).equals(requestDto.getCertificationNumber())) {
```
가독성을 위해 메소드로 분리해도 좋을 것 같은데 어떻게 생각하시나요??

kihyuk-jeong replied:
가독성을 위해 메서드 분리하고 추가로 멘토님께서 리뷰해 주신 것처럼 컨트롤러에서 boolean 값을 리턴 받는 것이 아닌, service 단에서 인증 불일치 시 Exception 처리하도록 변경하겠습니다.!

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/util/email/EmailConstants.java:
```diff
@@ -0,0 +1,8 @@
+public class EmailConstants {
+
+    public static final String TITLE = "Shoe-action 인증번호 안내";
+    public static final String FROM_ADDRESS = "rlgur405@gmail.com";
```
이런 값들은 데이터베이스에서 관리하거나 외부에서 주입해주는게 좋지 않을까요?

kihyuk-jeong replied:
이메일 같은 경우 노출되어도 크게 상관없다고 생각해서 상수로 정의하였는데, 개인 정보 관련된 것들은 외부 주입으로 변경하는 것이 보안상 더 옮은 방법 같습니다. 외부 주입으로 변경하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/util/email/EmailContentTemplate.java:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    public String parse() {
+        return prefix.concat(certificationNumber).concat(suffix);
```
`String.fotmat()`을 사용하는 것도 좋습니다~

kihyuk-jeong replied:
`자바의 신` 책에서 `concat`메서드는 항상 `new String`으로 객체를 생성해서 메모리 문제 때문에 거의 사용하지 않는다고 봤었는데 급히 작성하느라 신경 쓰지 못한 것 같습니다. 변경하겠습니다~

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/util/email/EmailContentTemplate.java:
```diff
@@ -0,0 +1,16 @@
+public class EmailContentTemplate {
+    private final String prefix = "[Shoe-Auction] 인증번호는 ";
+    private final String suffix = "입니다.";
+    private String certificationNumber;
```
이 값은 굳이 상태로 두지 않고, 문자열을 빌드할때 인자로 받아도 될 것 같습니다~

kihyuk-jeong replied:
넵 인증번호를 파라미터로 받아서 이메일 내용 객체를 리턴하는 방식으로 변경하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -86,5 +93,32 @@ public ResponseEntity logout() {
 
+    //비밀번호 찾기 : 이메일 입력시, 존재하는 이메일이면 휴대폰인증과 이메일인증 중 택1 하도록 구현
+    //휴대폰 인증 선택시 : sendSms / SmsVerification 핸들러
+    //이메일 인증 선택시 : sendEmail / emailVerification 핸들러
```
`//` 뒤에 띄어쓰기 하나 써주시면 가독성이 좀 더 좋아질 것 같네요~

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -86,5 +93,32 @@ public ResponseEntity logout() {
+    //비밀번호 찾기 : 이메일 입력시, 존재하는 이메일이면 휴대폰인증과 이메일인증 중 택1 하도록 구현
+    //휴대폰 인증 선택시 : sendSms / SmsVerification 핸들러
+    //이메일 인증 선택시 : sendEmail / emailVerification 핸들러
+    @GetMapping("/find/password/{email}")
```
이 URL은 리소스와 행위에 대해 잘 표현하고 있는걸까요?

kihyuk-jeong replied:
URL 네이밍 측면에서 고민을 오래 하다가 자주 방문하는 사이트에서 find라는 URL을 따라 썼었는데 해당 핸들러의 행위 자체가 휴대폰 번호와 이메일을 응답받기 위한 핸들러기 때문에 `password`라는 URL은 옳지 않은 것 같습니다. 조금 더 고민해서 수정하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -86,5 +93,32 @@ public ResponseEntity logout() {
 
+    @PostMapping("/email-certification/confirms")
+    public ResponseEntity emailVerification(@RequestBody EmailCertificationRequest requestDto) {
+        if(!emailCertificationService.verifyEmail(requestDto)) {
```
이런 validation 같은 경우는 예외를 던지게 만들어도 됩니다. 그리고 Exception Handler에서 해당 예외를 처리하게하면 좀 더 코드가 간결해집니다~

kihyuk-jeong replied:
서비스에서 예외 처리하는 것이 더 깔끔한 코드가 될 것 같습니다. 리뷰 반영하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -86,5 +93,32 @@ public ResponseEntity logout() {
+        return OK;
+    }
+
+    @PatchMapping("password-nonLogin")
```
이 URL을 통해 유추할 수 있는 행위는 무엇일까요?

kihyuk-jeong replied:
비밀번호 찾기에서 휴대폰 또는 이메일 인증이 완료된 사용자의 이메일을 변경하는 URL인데, 로그인 된 상태로 마이페이지에서 비밀번호를 변경하는 URL과 요청받는 리소스가 달라서 차별화를 두기 위해 `nonLogin`이라는 이름을 사용했습니다. 적용하면서도 많이 고민한 부분이라 URL 네이밍에 대해서 조금 더 고민해 봐야 할 것 같습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/EmailCertificationService.java:
```diff
@@ -0,0 +1,60 @@
+
+        emailCertificationDao.createEmailCertification(email, randomNumber);
+
+        log.info(emailCertificationDao.getEmailCertification(email));
```
로그 레벨을 info로 정하신 이유가 있으신가요?

kihyuk-jeong replied:
인증번호 저장 여부를 테스트하기 위해 찍어놓은 로그인데 개발 단계에서 불필요한 것 같습니다. 삭제하겠습니다!

kihyuk-jeong committed:
refactor : 리뷰 반영

1. snake case로 정의한 메서드명 변경
2. EmailConstants에 선언한 TITLE 상수 오타 수정
3. ResponseConstants에 선언한 http status code 명확한 의미 부여(이름 수정)
4. 공통적으로 사용되는 난수 생성 메소드 utils로 분리
5. 응답용 Dto에 알맞게 dto 이름 변경
6. EmailCertificationService에 인증번호 확인하는 verifyEmail() 메소드, 가독성을 위해 세부적으로 분리
7. 이메일(개인정보) 외부주입으로 변경
8. concat -> String.format으로 변경
9. 이메일 내용을 생성하는 객체 EmailContentTemplate 수정
10. 이메일 인증번호 확인 핸들러인 emailVerification() 내부 로직 수정
11. 불필요한 로그 삭제

<issue-number>
 #15 : 비밀번호 찾기

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/certification/RandomNumberGeneration.java:
```diff
@@ -1,12 +1,10 @@
-public class UserConstants {
-    public static final String USER_ID = "email";
-
+public class RandomNumberGeneration {
```
`Generation`보단 `Generator` 가 나아보이네요~

kihyuk-jeong replied:
생성기 라는 이름이 더 알맞은 것 같네요. 감사합니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/exception/GlobalExceptionHandler.java:
```diff
@@ -55,4 +58,12 @@
+    @ExceptionHandler(AuthenticationNumberMismatchException.class)
+    public final ResponseEntity<Void> handleAuthenticationNumberMismatchException(
+        AuthenticationNumberMismatchException ex, WebRequest request) {
+        log.error("인증번호가 일치하지 않습니다 :  {}, detection time={} ", request.getDescription(false),
```
사용자가 인증번호를 잘못 입력한것을 error 정도의 레벨까지 올려야할까요?
잘못 입력한 것은 error 정도로 치명적인 수준은 아닌 것 같습니다~

kihyuk-jeong replied:
단순히 사용자가 인증번호를 잘못 입력한 것이기 때문에 그에 대한 정보를 남기는 info 레벨로 내려서 처리하는 것이 맞는 것 같습니다. 
로그인 실패의 경우에도 사용자가 id/pw를 잘못 입력한 경우이기 때문에 log level을 내리는게 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/EmailCertificationService.java:
```diff
@@ -18,11 +19,13 @@
 public class EmailCertificationService {
 
+    @Value("${external.email}")
+    private String FROM_ADDRESS;
```
여기는 카멜케이스를 지키지 않은 이유가 있으실까요?

kihyuk-jeong replied:
기존에는 상수로 사용하고 있어서 변수와 구분하는 차원에서 `snake case`를 사용했는데 변수를 외부 주입받는 방법으로 변경했기 때문에 `camel case`로 변경하겠습니다.

f-lab-dev replied:
보통 대문자를 사용하는건 static에 주로 사용합니다~

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/EmailCertificationService.java:
```diff
@@ -18,11 +19,13 @@
 @Service
 public class EmailCertificationService {
 
+    @Value("${external.email}")
```
이런 property도 구조를 잘 정해야합니다. 이 `external.email` 이라는 이름을 보고 추후에 이 소스코드를 유지보수하는 사람이 이 값의 의미를 제대로 파악할 수 있을까요?

kihyuk-jeong replied:
조금 더 명확한 의미를 가질 수 있는 이름을 부여하겠습니다.
추가적으로 `@Value`의 타입 안정성 문제와 추후 외부 주입 추가 사용시 동일한 어노테이션의 추가로 프로덕션 코드가 오염될 가능성이 있다고 생각됩니다. 따라서  `@ConfigurationProperties` 로 대체하겠습니다.

kihyuk-jeong committed:
refactor : 리뷰반영[2]

1. 외부 주입 방식 변경 (@Value -> @ConfigurationProperties)
2. 변수명 수정(snake case -> camel case)
3. 로그 레벨 down (로그인 실패/인증번호 불일치 error -> info)
4. 난수 생성 util 클래스 이름 변경
5. EmailCertificationService에 적용한 리뷰 반영 내용 SmsCertificationService에 동일하게 적용

<issue-number>
    #15 : 비밀번호 찾기

kihyuk-jeong committed:
test : 비밀번호 찾기 관련 테스트코드 추가

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/certification/coolSms/SmsMessageTemplate.java:
```diff
@@ -3,14 +3,9 @@
-    public String parse() {
-        return prefix.concat(certificationNumber).concat(suffix);
+    public String parse(String certificationNumber) {
+        return String.format("%s%s%s", prefix,certificationNumber,suffix);
```
어차피 템플릿이 정해졌다면 위의 상수들을 따로 선언할 필요 없이 하나의 포맷으로 합쳐도 되지 않을까요?

ROMANIAPEOPLE replied:
추후 이메일 내용이 변경될 가능성이 있다고 판단하여 템플릿을 만들어 사용했는데 굳이 상수까지 선언할 필요는 없을 것 같습니다!  리뷰 반영하겠습니다.

kihyuk-jeong replied:
추후 거래 관련 내용의 메세지도  동일한 Template 클래스에서 사용할 목적으로 만들었는데 상수들을 선언하면 오히려 확장의 유연성이 떨어질 것 같네요!  리뷰 반영하겠습니다. 
또한 추후 확장과 관련해서 `parse`라는 메소드 네임보다 인증과 관련된 내용이라는 것을 명시해줄 수 있도록 메소드 이름 변경하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/exception/GlobalExceptionHandler.java:
```diff
@@ -44,9 +44,8 @@
-        log.error("Failed to signUp ::  {}, detection time={} ", request.getDescription(false),
-            LocalDateTime.now(), ex);
+        UserNotFoundException ex) {
+        log.info("로그인 실패 : 존재하지 않는 ID 또는 패스워드 불일치",ex);
```
이것 또한 info로 찍어두면 너무 실서버에서 과도한 로그가 생성되지 않을까요?
비밀번호를 틀린다거나 인증번호를 틀리는 일은 흔한 일이고 그렇게 크리티컬한 일이 아니긴 합니다.
다만 여러번 연속으로 틀리면 무작위 대입 공격일 수도 있으니 막아야겠죠~

kihyuk-jeong replied:
해당 로그는 개발 환경에서만 사용하고 실제 배포 환경에서는 불필요한 로그 같습니다. `debug` 레벨로 하향시킨 후, 개발 단계에서만 사용하는 방향으로 설정하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/LoginService.java:
```diff
@@ -43,6 +44,6 @@ public String getLoginUser() {
 
         return userRepository.findByEmail(email)
-            .orElseThrow().toUserInfoDto();
+            .orElseThrow(() -> new UnauthenticatedUserException("error : 인증되지 않은 사용자")).toUserInfoDto();
```
email로 찾았는데 없는거면 `존재하지 않는 사용자`가 아닐까요?
만약 이 예외를 캐치하는 개발자가 이 메세지를 보면 어떻게 받아들일까요?

kihyuk-jeong replied:
이부분은 컨트롤러 메서드에서 `@LoginCheck`를 통해 로그인 여부를 검사 후 실행하는 로직이기 때문에 실제로 해당 예외가 사용자가 존재하지 않아서 발생할 일은 없다고 생각하고,  간혹 예외적인 상황으로 로그인이 풀릴 경우를 생각해서  처리한 예외 메세지 입니다!

f-lab-dev replied:
Service는 Controller의 존재에 대해 몰라야합니다. 이것은 그럼 Controller를 의식하면서 만든 코드가 아닐까요?

kihyuk-jeong replied:
해당 메소드가 저장된 사용자의 정보를 가져오는 로직이기때문에  존재하지 않는 사용자로 처리하는것이 맞는 것 같습니다. 수정하겠습니다.!

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/SmsCertificationService.java:
```diff
@@ -1,47 +1,44 @@
 @RequiredArgsConstructor
 @Service
+@Setter
+@ConfigurationProperties("external")
```
여기서도 `external`은 너무 범위가 넓은 단어네요

kihyuk-jeong replied:
인증과 관련된 상수들 이라는 뜻의 좁은 범위로 수정하겠습니다.

kihyuk-jeong committed:
refactor : 리뷰반영

1.property 구조 이름 변경
2. log 레벨 하향 조정
3.SmsMessageTemplate, EmailContentTemplate 수정

kihyuk-jeong committed:
chore : 오타 수정

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/certification/coolSms/SmsMessageTemplate.java:
```diff
@@ -1,11 +1,8 @@
-
-    public String parse(String certificationNumber) {
-        return String.format("%s%s%s", prefix,certificationNumber,suffix);
+    public String getCertificationContent(String certificationNumber) {
```
`get`도 좋지만 `build`라는 단어를 써도 좋을 것 같네요~

kihyuk-jeong replied:
항상 네이밍 부분에서 고민을 많이 하는데 content 내용을 만드는 작업이기 때문에 build를 사용하는 것이 전달이 더 잘 될 것 같습니다. 리뷰 반영하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/certification/coolSms/SmsMessageTemplate.java:
```diff
@@ -1,11 +1,8 @@
-    public String parse(String certificationNumber) {
-        return String.format("%s%s%s", prefix,certificationNumber,suffix);
+    public String getCertificationContent(String certificationNumber) {
+        return String.format("%s%s%s", "[Shoe-Auction] 인증번호는 ",certificationNumber,"입니다.");
```
제 뜻은 `"%s%s%s"` 이거를 포맷으로 만들라는 말이였습니다~

kihyuk-jeong committed:
refactor : 리뷰반영

1. 기존 foramt을 사용하던 로직 StringBuilder로 변경
2. 메소드 이름 수정 (get -> build)
3. 예외 메시지 수정

aimiebell commented on src/main/java/com/flab/shoeauction/common/utils/certification/coolSms/coolSmsConstants.java:
```diff
@@ -0,0 +1,6 @@
+package com.flab.shoeauction.common.utils.certification.coolSms;
+
+public class coolSmsConstants {
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Class names should start with an upper case letter (NM_CLASS_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-class-names-should-start-with-an-upper-case-letter-nm-class-naming-convention).
Class names should be nouns, in mixed case with the first letter of each internal word capitalized. Try to keep your class names simple and descriptive. Use whole words-avoid acronyms and abbreviations (unless the abbreviation is much more widely used than the long form, such as URL or HTML).

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/certification/coolSms/SmsMessageTemplate.java:
```diff
@@ -1,8 +1,14 @@
-        return String.format("%s%s%s", "[Shoe-Auction] 인증번호는 ",certificationNumber,"입니다.");
+    public String buildCertificationContents(String certificationNumber) {
+
+        StringBuilder bd = new StringBuilder();
```
`bd`로 줄이는건 흔하지 않은 네이밍이네요~

kihyuk-jeong committed:
refactor : 리뷰 반영
